src/transforms/dedupe.rs

```diff
@@ -0,0 +1,414 @@
+        Ok(Box::new(Dedupe::new(
+            self.cache.num_entries,
+            self.fields.clone(),
+        )))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "dedupe"
+    }
+}
+
+/**
+* A CacheEntry comes in two forms, depending on the FieldMatchConfig in use.
+*
+* When matching fields, a CacheEntry contains a vector of optional 2-tuples.  Each element in the
+* vector represents one field in the corresponding LogEvent.  Elements in the vector will correspond
+* 1:1 (and in order) to the fields specified in "fields.match".  The tuples each store the
+* TypeId for this field and the data as Bytes for the field.  There is no need to store the field
+* name because the elements of the vector correspond 1:1 to "fields.match", so there is never any
+* ambiguity about what field is being referred to.  If a field from "fields.match" does not show
+* up in an incoming Event, the CacheEntry will have None in the correspond location in the vector.
+*
+* When ignoring fields, a CacheEntry contains a vector of 3-tuples.  Each element in the vector
+* represents one field in the corresponding LogEvent.  The tuples will each contain the field name,
+* TypeId, and data as Bytes for the corresponding field (in that order).  Since the set of fields
+* that might go into CacheEntries is not known at startup, we must store the field names as part of
+* CacheEntries.  Since Event objects store their field in alphabetic order (as they are backed by a
+* BTreeMap), and we build CacheEntries by iterating over the fields of the incoming Events, we know
+* that the CacheEntries for 2 equivalent events will always contain the fields in the same order.
+**/
+#[derive(PartialEq, Eq, Hash)]
+enum CacheEntry {
+    Match(Vec<Option<(TypeId, Bytes)>>),
+    Ignore(Vec<(Atom, TypeId, Bytes)>),
+}
+
+struct Null;
+
+fn type_id_for_value(val: &Value) -> TypeId {
+    match val {
+        Value::Bytes(_) => TypeId::of::<Bytes>(),
+        Value::Timestamp(_) => TypeId::of::<DateTime<Utc>>(),
+        Value::Integer(_) => TypeId::of::<i64>(),
+        Value::Float(_) => TypeId::of::<f64>(),
+        Value::Boolean(_) => TypeId::of::<bool>(),
+        Value::Map(_) => TypeId::of::<BTreeMap<Atom, Value>>(),
+        Value::Array(_) => TypeId::of::<Vec<Value>>(),
+        // There's no real type in Rust for Null, so we make a new type just for getting a unique
+        // TypeId. All that matters is that all possible Value types have a distinct TypeId.
+        Value::Null => TypeId::of::<Null>(),
+    }
+}
+
+impl Dedupe {
+    pub fn new(num_entries: usize, field_match_config: FieldMatchConfig) -> Self {
+        Self {
+            config: DedupeConfig {
+                fields: field_match_config,
+                cache: CacheConfig { num_entries },
```

Comment:
Probably would have been better to have this fn accept a `CacheConfig` so that later when we want more than just `num_entries` in the config we don't need to break API. :)

Reply:
Good call.  Done!